### ISSUE TITLE
i18n: bugs and inconsistencies in date-handling

### ISSUE THREAD

**Issue by gschintgen** [johnstray/gs-blog#80]
1st issue:
When adding a blog post without manually specifying a "publish date", gs-blog automatically saves the current date and time, but incorrectly (IMO): It is stored in the way the current language file specifies the date format. If it's d/m/Y instead of m/d/Y that's how the date is saved. (It should always be stored independently of the language setting, e.g. utc timestamp or in some other fixed day month order.)
When I have a look at such a post's options it shows in the explanatory text that the format is mm/dd/yy but the string in the text box is actually dd/mm/yy.

So one "solution" would be for the translators to adapt the corresponding text strings, *but* that's not going to work: The blog's code expects the date to be in mm/dd/yy. Blog posts which have 25/03/2016 will have no date on the blog page! The date is only shown if the admin interface shows the date as 03/25/2016 even though dd/mm/yy is specified in the language file (fr_FR in this case).

2nd issue:
If a date is shown by gs-blog, it's always in English! ("March 25th, 2016")

Note that for testing purposes I switched multiple times from English to German to French etc. But this should not break date handling.

**Comment by johnstray** [comment]
What I might to fix this is require that an ISO date format be used when creating a post and for storing that date in file, or converting the date to an ISO format if another date format is used.

As for the display of the date, the layout files use the date() function taking 'DATE_DISPLAY' from the language file as the format. I suspect, however, that the language file doesn't get loaded on the front end and that's why they are always showing in English. I'll have to find another way of doing this I think. 
A workaround for now is to manually specify a date format for the date() function in the layout-* files.

I'll look into this further over the next couple of days.

**Comment by johnstray** [comment]
Upon further investigation, the date show on the front end does respect the currently active language file, so this is not an issue.

I have however made the following changes:
- The date is now stored in file using the MySQL date format (Y-m-d H:i:s).
- When creating/editing a post, any date format supported by strtotime() is now supported, including "today", etc.

**Comment by gschintgen** [comment]
> Upon further investigation, the date show on the front end does respect the currently active language file, so this is not an issue.

Sorry, I'll have to disagree. I hope it's okay with you that I'm reopening this issue.

On my test installation the date is *not* translated. I had a look at the PHP manual and it states that the date() function always prints the date in English. This may be insignificant if only numeric values are used, but it does make a difference if the full month is part of the date string.

According to the manual:
> To format dates in other languages, you should use the setlocale() and strftime() functions instead of date().

Such a method seems to be implemented already in the Blog class: get_locale_date(). It's also available as formatPostDate() in inc/common.php. I'm noticing just now that I even translated the corresponding "Help" strings :-)

I have not yet tried adapting the layout files.

**Comment by gschintgen** [comment]
I had a look at the code and here are my findings:
- formatPostDate is deprecated (according to the comments in common.php), but still documented in the "Help" section.
- formatPostDate uses strftime() which is the way to go for properly localized dates (at leas as far as I can tell based on the php manual), so it should probably be revived.
- formatPostDate uses a hardcoded format string.
Here's my proposal: modify formatPostDate to use a format string stored in the language files and use this function to print the date everywhere (with differing format strings as is the case right now)

**Comment by johnstray** [comment]
I think I understand the issue a bit better now.
I've had a bit of a play around, taking in the suggestions that you made, and have come up with a solution. To implement this though is going to require some minor changes to the language files.

So, this is what I did:
- In `$Blog->get_locale_date()` I changed `setlocale()` so that it will take an array of language code formats from the active language file. This was required because according to the manual for setlocale(), different strings are required depending on different systems, but this function allows for an array of possible strings to try until success.
- `common.php formatPostDate()` will now respect the 'DATE_DISPLAY' format string.
- In the language files: 
  - 'LANGUAGE_CODE' is now an array of possible language codes supported by setlocale()
  - Date formats have been converted to a style that is supported by strftime() - This is different to date()

Still to do:
- Find any spot where a date is displayed and modify it to use the reborn formatPostDate() function.
